include/rapidcheck/state/gen/ParallelCommands.hpp

```diff
@@ -0,0 +1,341 @@
+#pragma once
+
+#include "rapidcheck/state/gen/Commands.h"
+
+namespace rc {
+namespace state {
+namespace gen {
+
+namespace detail {
+template <typename Cmd, typename GenFunc>
+class ParallelCommandsGen {
+public:
+  using CmdSP = std::shared_ptr<const Cmd>;
+  using Model = typename Cmd::Model;
+  using Sut = typename Cmd::Sut;
+
+  template <typename ModelArg, typename GenFuncArg>
+  ParallelCommandsGen(ModelArg &&initialState, GenFuncArg &&genFunc)
+      : m_initialState(std::forward<ModelArg>(initialState))
+      , m_genFunc(std::forward<GenFuncArg>(genFunc)) {}
+
+  Shrinkable<ParallelCommands<Cmd>> operator()(const Random &random,
+                                               int size) const {
+    return generateCommands(random, size);
+  }
+
+private:
+  struct CommandEntry {
+    CommandEntry(Random &&aRandom, Shrinkable<CmdSP> &&aShrinkable)
+        : random(std::move(aRandom))
+        , shrinkable(std::move(aShrinkable)) {}
+
+    Random random;
+    Shrinkable<CmdSP> shrinkable;
+  };
+
+  struct CommandSequence {
+    CommandSequence(const GenFunc &func, int sz)
+        : genFunc(func)
+        , size(sz) {}
+
+    Model initialState;
+    GenFunc genFunc;
+    int size;
+    std::vector<CommandEntry> entries;
+
+
+    void repairEntriesFrom(std::size_t start) {
+      for (auto i = start; i < entries.size(); i++) {
+        if (!repairEntryAt(i)) {
+          entries.erase(begin(entries) + i--);
+        }
+      }
+    }
+
+    bool repairEntryAt(std::size_t i) {
+      using namespace ::rc::detail;
+      try {
+        auto &entry = entries[i];
+        const auto cmd = entry.shrinkable.value();
```

Comment:
Since you don't have the `preState` here, what are you actually expecting will happen here? The `repairEntriesFrom` method was intended to update the `preState` and if that failed (i.e. the command had become invalid) it would be regenerated.

This just retrieves the command which in an of itself will always succeed.